File: python/main/topjslibs/top_js_libraries.py
import asyncio
from collections import OrderedDict

from topjslibs.google_search_result import GoogleSearchResult
from topjslibs.page_js_sources import PageJsSources

SEARCH_URL = "https://www.google.com/search?q="


class TopJsLibraries:

    def __init__(self, links_limit, top_limit, fetch_html=lambda url: ""):
        self.links_limit = links_limit
        self.top_limit = top_limit
        self.fetch_html = fetch_html

    async def __get_page_js_sources(self, url) -> set[str]:
        return PageJsSources(fetch_html=lambda: self.fetch_html(url)).get()

    async def count(self, search_str) -> OrderedDict[dict[str, int]]:
        links = GoogleSearchResult(links_limit=self.links_limit,
                                   fetch_html=lambda: self.fetch_html(
                                           SEARCH_URL + search_str)).get_links()
        get_page_js_sources_tasks = [
            asyncio.create_task(self.__get_page_js_sources(link)) for link in
            links]

        pages_js_sources = await asyncio.gather(*get_page_js_sources_tasks)
        all_js_sources = [js for sublist in pages_js_sources for js in sublist]
        js_count = {}

        for js in all_js_sources:
            js_count.setdefault(js, 0)
            js_count[js] += 1

        js_count_sorted = sorted(js_count.items(),
                                 key=lambda item: (-item[1], item[0]))

        return OrderedDict(js_count_sorted[:self.top_limit])

File: nodejs/src/topjslibs/TopJsLibraries.ts
import {GoogleSearchResult} from "./GoogleSearchResult";
import {PageJsSources} from "./PageJsSources";

export class TopJsLibraries {

  private static readonly SEARCH_URL = "https://www.google.com/search?q=";

  private readonly linksLimit: number;
  private readonly topLimit: number;
  private readonly fetchHtml: (url: string) => Promise<string>;

  constructor(readonly linksLimitValue: number,
              readonly topLimitValue: number,
              readonly fetchHtmlLambda: (url: string) => Promise<string>) {
    this.linksLimit = linksLimitValue;
    this.topLimit = topLimitValue;
    this.fetchHtml = fetchHtmlLambda;
  }

  public async count(searchStr: string): Promise<Array<[string, number]>> {
    const fetchSearchResultHtml = () => this.fetchHtml(TopJsLibraries.SEARCH_URL + searchStr);
    const links = await new GoogleSearchResult(this.linksLimit, fetchSearchResultHtml).getLinks();
    const pageJsSources = new Array<Promise<Set<string>>>();

    for (const link of links) {
      pageJsSources.push(this.getPageJsSources(link));
    }

    const jsCount = new Map<string, number>();
    const pageJsSourcesResults = await Promise.allSettled(pageJsSources);

    for (const pageJsSourcesResult of pageJsSourcesResults) {
      if (pageJsSourcesResult.status === "rejected") {
        continue;
      }

      const jsSources = (pageJsSourcesResult as PromiseFulfilledResult<Set<string>>).value;

      for (const jsSource of jsSources) {
        let counter = 0;

        if (jsCount.has(jsSource)) {
          counter = jsCount.get(jsSource)!;
        }

        jsCount.set(jsSource, counter + 1);
      }
    }

    const sortedJsCount = new Array<[string, number]>();
    for (const entry of jsCount.entries()) {
      sortedJsCount.push(entry);
    }

    sortedJsCount.sort((a, b) => (a[1] > b[1] || (a[1] === b[1] && a[0] < b[0])) ? -1 : 1);

    return sortedJsCount.slice(0, this.topLimit);
  }

  private async getPageJsSources(url: string): Promise<Set<string>> {
    return new PageJsSources(() => this.fetchHtml(url)).get();
  }
}
